File: 2384-largest-palindromic-number/2384-largest-palindromic-number.c
char* largestPalindromic(char* num) {
    int len = strlen(num);
    int freq[10] = {0, };
    int valid_num = 0;
    int left = len / 2 - 1, mid = len / 2, right = len / 2;
    
    for (int i = 0; i < len; i++) {
        freq[num[i]-'0']++;
    }
    
    for (int i = 0; i < 10; i++) {
        if (freq[i]) valid_num++;
    }
    
    if (valid_num == 1 && freq[0]) return "0";
    
    char* new_num = (char*) malloc(sizeof(char) * (len+1));
    for (int i = 9; i >= 0; i--) {
        if (freq[i] % 2) {
            new_num[mid] = i + '0';
            freq[i]--;
            right++;
            break;
        }
    }
    
    for (int i = 0; i < 10; i++) {
        while (freq[i] > 1) {
            new_num[left--] = i + '0';
            new_num[right++] = i + '0';
            freq[i] -= 2;
        }
    }
    
    while (new_num[left+1] == '0') {
        left++; right--;
    }
    
    new_num[right] = '\0';
    return new_num+left+1;    
}

File: 2384-largest-palindromic-number/2384-largest-palindromic-number.py
class Solution:
    def largestPalindromic(self, num: str) -> str:
        numDict = {}
        for c in num:
            n = int(c)
            if n not in numDict.keys():
                numDict[n] = 1
            else: numDict[n] += 1
        val = ""
        for n in range(9, -1, -1):
            if n in numDict.keys() and numDict[n] % 2:
                val = str(n)
                numDict[n] -= 1
                break
            
        for n in range(0, 10):
            if n in numDict.keys():
                c = str(n)
                while numDict[n] > 1:
                    numDict[n] -= 2
                    val = c + val + c

        while len(val) > 2 and val[-1] == '0':
            val = val[1:-1]
        if len(val) == 2 and val[-1] == '0':
            val = val[0]
            
        return val
        
            
        